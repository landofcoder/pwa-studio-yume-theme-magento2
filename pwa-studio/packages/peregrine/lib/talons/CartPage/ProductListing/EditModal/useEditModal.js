import { useState } from 'react';

import { useAppContext } from '../../../../context/app';

/**
 * This talon contains logic for a product edit modal used on a cart page.
 * It returns prop data for rendering an interactive modal component.
 *
 * @function
 *
 * @return {EditModalTalonProps}
 *
 * @example <caption>Importing into your project</caption>
 * import { useEditModal } from '@magento/peregrine/lib/talons/CartPage/ProductListing/EditModal/useEditModal';
 */
export const useEditModal = () => {
    const [{ drawer }, { closeDrawer }] = useAppContext();
    const isOpen = drawer === 'product.edit';

    const [variantPrice, setVariantPrice] = useState(null);

    return {
        handleClose: closeDrawer,
        isOpen,
        setVariantPrice,
        variantPrice
    };
};

/** JSDocs type definitions */

/**
 * Object type returned by the {@link useEditModal} talon.
 * It provides props data for rendering an edit modal component.
 *
 * @typedef {Object} EditModalTalonProps
 *
 * @property {function} handleClose Callback function for handling the closing event of the modal.
 * @property {boolean} isOpen True if the modal is open. False otherwise.
 * @property {function} setVariantPrice Function for setting a product's variant price.
 * @property {Object} variantPrice The variant price for a product. See [Money object]{@link https://devdocs.magento.com/guides/v2.4/graphql/product/product-interface.html#Money}.
 */
