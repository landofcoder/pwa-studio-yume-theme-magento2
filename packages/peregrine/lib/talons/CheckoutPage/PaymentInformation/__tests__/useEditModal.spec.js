import React from 'react';

import createTestInstance from '../../../../util/createTestInstance';
import { useEditModal } from '../useEditModal';

jest.mock('@apollo/client', () => {
    return {
        useQuery: jest.fn().mockReturnValue({
            data: {
                cart: {
                    selected_payment_method: { code: 'braintree' }
                }
            }
        })
    };
});

jest.mock('../../../../context/cart', () => ({
    useCartContext: jest.fn().mockReturnValue([{ cartId: '123' }])
}));

const Component = props => {
    const talonProps = useEditModal(props);

    return <i talonProps={talonProps} />;
};

const getTalonProps = props => {
    const tree = createTestInstance(<Component {...props} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    const update = newProps => {
        tree.update(<Component {...newProps} />);

        return talonProps;
    };

    return { talonProps, tree, update };
};

test('Should return correct shape', () => {
    const { talonProps } = getTalonProps({
        onClose: () => {},
        queries: { getSelectedPaymentMethodQuery: '' }
    });

    expect(talonProps).toMatchSnapshot();
});

test('Should call onClose when handleClose is called', () => {
    const onClose = jest.fn();

    const { talonProps } = getTalonProps({
        onClose,
        queries: { getSelectedPaymentMethodQuery: '' }
    });
    const { handleClose } = talonProps;

    handleClose();

    expect(onClose).toHaveBeenCalled();
});

test('Should call onClose when handlePaymentSuccess is called', () => {
    const onClose = jest.fn();

    const { talonProps } = getTalonProps({
        onClose,
        queries: { getSelectedPaymentMethodQuery: '' }
    });
    const { handlePaymentSuccess } = talonProps;

    handlePaymentSuccess();

    expect(onClose).toHaveBeenCalled();
});

test('Should set updateButtonClicked to true when handleUpdate is called', () => {
    const { talonProps, tree } = getTalonProps({
        onClose: () => {},
        queries: { getSelectedPaymentMethodQuery: '' }
    });

    talonProps.handleUpdate();

    expect(
        tree.root.findByType('i').props.talonProps.updateButtonClicked
    ).toBeTruthy();
});

test('Should set isLoading to false when handleDropinReady is called', () => {
    const { talonProps, tree } = getTalonProps({
        onClose: () => {},
        queries: { getSelectedPaymentMethodQuery: '' }
    });

    talonProps.handleDropinReady();

    expect(tree.root.findByType('i').props.talonProps.isLoading).toBeFalsy();
});
