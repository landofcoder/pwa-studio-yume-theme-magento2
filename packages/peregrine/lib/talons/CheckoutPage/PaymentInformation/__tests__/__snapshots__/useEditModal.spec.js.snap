// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return correct shape 1`] = `
Object {
  "handleClose": [Function],
  "handleDropinReady": [Function],
  "handlePaymentError": [Function],
  "handlePaymentSuccess": [Function],
  "handleUpdate": [Function],
  "isLoading": true,
  "resetUpdateButtonClicked": [Function],
  "selectedPaymentMethod": "braintree",
  "updateButtonClicked": false,
}
`;
